synapse/python_dependencies.py

```diff
@@ -70,7 +70,7 @@
     "pyasn1>=0.1.9",
     "pyasn1-modules>=0.0.7",
     "bcrypt>=3.1.0",
-    "pillow>=4.3.0",
```

Comment:
I think this has come up before, but I was a bit surprised to see this. Broadly speaking, a version of Pillow that successfully ran on 3.6 should also successfully run on 3.7, 3.8, 3.9... etc. (There are deprecations and removals between minor Python releases, but these are uncommon.)  The pain is that Pillow uses a C extension and therefore needs binaries to be built for each interpreter version.

As you pointed out, olddeps was failing here: https://github.com/matrix-org/synapse/runs/4745589267?check_suite_focus=true

For completeness, I think what's happening here is:
- we used to be able to pull a binary wheel from PyPI on bionic. [See e.g.](https://github.com/matrix-org/synapse/runs/4840259529?check_suite_focus=true#step:4:1212)
  > 2022-01-17T11:47:41.2123816Z Collecting pillow==4.3.0
  > 2022-01-17T11:47:41.2318857Z   Downloading Pillow-4.3.0-cp36-cp36m-manylinux1_x86_64.whl (5.8 MB)
- Focal uses Python 3.8.
- Pillow doesn't "support" Python 3.8 [until Pillow 6.2.1](https://pillow.readthedocs.io/en/stable/installation.html#python-support). I think "support" means "provide binary wheels for".
- We still want debian oldstable to use its version, 5.4.1, so don't want to bump the min version to 6.2.1.
- For some reason, `pip/tox` in the olddeps job chooses Pillow 5.4.0 exactly. This doesn't have a Python 3.8 wheel.
- Therefore we need to build the C extension ourselves from scratch. And that's also why we need `libjpeg-devel`, for instance.